File: scripts/ingest/bls.ts
import { z } from 'zod';
import type { MetricId, Observation, Unit } from '@dmv/shared';
import { fetchWithRetry } from '../lib/http.js';
import { log } from '../lib/log.js';
import { IngestError } from '../lib/errors.js';
import { DMV_COUNTIES } from '../lib/counties.js';
import type { DataSource } from './DataSource.js';

const BLS_API_URL = 'https://api.bls.gov/publicAPI/v2/timeseries/data/';
const START_YEAR = '2015';
const END_YEAR = '2026';

const MSA_FEDERAL_SERIES = 'SMU11479009091000001';
const MSA_FEDERAL_FIPS = '11-metro';

const BlsDataPointSchema = z.object({
  year: z.string(),
  period: z.string(),
  value: z.string(),
  footnotes: z.array(z.unknown()).optional(),
});

const BlsSeriesSchema = z.object({
  seriesID: z.string(),
  data: z.array(BlsDataPointSchema),
});

const BlsResponseSchema = z.object({
  status: z.string(),
  message: z.array(z.string()).optional(),
  Results: z
    .object({
      series: z.array(BlsSeriesSchema),
    })
    .optional(),
});

interface SeriesMeta {
  fips: string;
  metric: MetricId;
  unit: Unit;
}

/** Convert BLS year + period code to ISO date string, or null for M13 (annual average). */
export function periodToIso(year: string, period: string): string | null {
  if (period === 'M13') return null;
  const month = period.slice(1).padStart(2, '0');
  return `${year}-${month}-01`;
}

function buildSeriesMeta(): Map<string, SeriesMeta> {
  const meta = new Map<string, SeriesMeta>();
  for (const c of DMV_COUNTIES) {
    meta.set(`LAUCN${c.fips}0000000003`, {
      fips: c.fips,
      metric: 'unemployment_rate',
      unit: 'percent',
    });
  }
  meta.set(MSA_FEDERAL_SERIES, {
    fips: MSA_FEDERAL_FIPS,
    metric: 'federal_employment',
    unit: 'count',
  });
  return meta;
}

/**
 * Parse a raw BLS API response into Observations.
 * Exported for unit testing without HTTP mocking.
 * Logs a warn for each requested series absent from the response.
 */
export function parseBlsResponse(
  raw: unknown,
  seriesMeta: Map<string, SeriesMeta>,
): Observation[] {
  const parsed = BlsResponseSchema.parse(raw);

  if (parsed.status !== 'REQUEST_SUCCEEDED') {
    const msg = parsed.message?.join('; ') ?? 'unknown error';
    throw new IngestError(`BLS request failed: ${msg}`, { source: 'bls' });
  }

  const series = parsed.Results?.series ?? [];
  const receivedIds = new Set(series.map((s) => s.seriesID));

  for (const id of seriesMeta.keys()) {
    if (!receivedIds.has(id)) {
      log.warn({ series: id }, 'bls: series absent from response');
    }
  }

  const all: Observation[] = [];

  for (const s of series) {
    const meta = seriesMeta.get(s.seriesID);
    if (!meta) {
      log.warn({ series: s.seriesID }, 'bls: unexpected series in response; skipping');
      continue;
    }

    let seriesCount = 0;
    for (const point of s.data) {
      const observedAt = periodToIso(point.year, point.period);
      if (observedAt === null) continue;

      const num = Number(point.value);
      if (!Number.isFinite(num)) continue;

      all.push({
        source: 'bls',
        series: s.seriesID,
        fips: meta.fips,
        metric: meta.metric,
        observedAt,
        value: num,
        unit: meta.unit,
      });
      seriesCount++;
    }

    log.info({ series: s.seriesID, fips: meta.fips, count: seriesCount }, 'bls: parsed series');
  }

  return all;
}

function getApiKey(): string {
  const key = process.env.BLS_API_KEY;
  if (!key) {
    throw new IngestError('BLS_API_KEY not set in environment', { source: 'bls' });
  }
  return key;
}

export class BlsSource implements DataSource {
  readonly name = 'bls';
  readonly cadence = 'monthly' as const;

  async fetch(): Promise<Observation[]> {
    const registrationkey = getApiKey();
    const seriesMeta = buildSeriesMeta();
    const allSeriesIds = [...seriesMeta.keys()];

    log.info({ count: allSeriesIds.length }, 'bls: posting batch request');

    const body = JSON.stringify({
      seriesid: allSeriesIds,
      startyear: START_YEAR,
      endyear: END_YEAR,
      registrationkey,
    });

    let raw: unknown;
    try {
      const res = await fetchWithRetry(BLS_API_URL, {
        label: 'bls:batch',
        method: 'POST',
        body,
        headers: { 'Content-Type': 'application/json' },
      });
      raw = await res.json();
    } catch (err) {
      throw new IngestError('BLS batch request failed', { source: 'bls' }, err);
    }

    const observations = parseBlsResponse(raw, seriesMeta);
    log.info({ count: observations.length }, 'bls: fetch complete');
    return observations;
  }
}

File: go/internal/ingest/bls/bls.go
// Package bls ports scripts/ingest/bls.ts. One POST to the BLS public v2 API
// with all DMV county LAUS series plus the federal-employment MSA series.
// M13 annual averages are dropped; M01–M12 become observedAt = year-MM-01.
package bls

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"log/slog"
	"net/http"
	"strconv"
	"strings"

	"github.com/meowmix1337/dmv_housing_hub/go/internal/counties"
	httpclient "github.com/meowmix1337/dmv_housing_hub/go/internal/http"
	pkglog "github.com/meowmix1337/dmv_housing_hub/go/internal/log"
	"github.com/meowmix1337/dmv_housing_hub/go/internal/types"
)

const (
	apiURL           = "https://api.bls.gov/publicAPI/v2/timeseries/data/"
	startYear        = "2015"
	endYear          = "2026"
	msaFederalSeries = "SMU11479009091000001"
	msaFederalFIPS   = "11-metro"
	requestSucceeded = "REQUEST_SUCCEEDED"
)

type Config struct {
	APIKey string `env:"BLS_API_KEY,required"`
}

type seriesMeta struct {
	fips   string
	metric types.MetricId
	unit   types.Unit
}

type requestBody struct {
	SeriesID        []string `json:"seriesid"`
	StartYear       string   `json:"startyear"`
	EndYear         string   `json:"endyear"`
	RegistrationKey string   `json:"registrationkey"`
}

type dataPoint struct {
	Year   string `json:"year"`
	Period string `json:"period"`
	Value  string `json:"value"`
}

type seriesEntry struct {
	SeriesID string      `json:"seriesID"`
	Data     []dataPoint `json:"data"`
}

type apiResponse struct {
	Status  string   `json:"status"`
	Message []string `json:"message"`
	Results struct {
		Series []seriesEntry `json:"series"`
	} `json:"Results"`
}

// PeriodToISO converts a BLS year + period code to ISO date.
// Returns "" for M13 (annual average) which callers must skip.
func PeriodToISO(year, period string) string {
	if period == "M13" {
		return ""
	}
	month := strings.TrimPrefix(period, "M")
	if len(month) == 1 {
		month = "0" + month
	}
	return year + "-" + month + "-01"
}

func buildSeriesMeta() map[string]seriesMeta {
	m := make(map[string]seriesMeta, len(counties.All())+1)
	for _, c := range counties.All() {
		m["LAUCN"+c.FIPS+"0000000003"] = seriesMeta{
			fips:   c.FIPS,
			metric: types.MetricUnemploymentRate,
			unit:   types.UnitPercent,
		}
	}
	m[msaFederalSeries] = seriesMeta{
		fips:   msaFederalFIPS,
		metric: types.MetricFederalEmployment,
		unit:   types.UnitCount,
	}
	return m
}

type Source struct {
	cfg    Config
	client *httpclient.Client
	logger *slog.Logger
	apiURL string // override for tests
}

func New(cfg Config, client *httpclient.Client) *Source {
	return &Source{
		cfg:    cfg,
		client: client,
		logger: pkglog.Default(),
		apiURL: apiURL,
	}
}

func (s *Source) Name() string           { return "bls" }
func (s *Source) Cadence() types.Cadence { return types.CadenceMonthly }

// SetLogger overrides the package-default logger. Used by cmd/ingest-all
// to inject a per-source attribute on every record.
func (s *Source) SetLogger(l *slog.Logger) {
	if l != nil {
		s.logger = l
	}
}

func (s *Source) Fetch(ctx context.Context) ([]types.Observation, error) {
	meta := buildSeriesMeta()
	ids := make([]string, 0, len(meta))
	for id := range meta {
		ids = append(ids, id)
	}

	s.logger.Info("bls: posting batch request", "count", len(ids))

	body, err := json.Marshal(requestBody{
		SeriesID:        ids,
		StartYear:       startYear,
		EndYear:         endYear,
		RegistrationKey: s.cfg.APIKey,
	})
	if err != nil {
		return nil, fmt.Errorf("marshal bls request: %w", err)
	}

	req, err := http.NewRequestWithContext(ctx, http.MethodPost, s.apiURL, bytes.NewReader(body))
	if err != nil {
		return nil, fmt.Errorf("new bls request: %w", err)
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := s.client.Do(ctx, req)
	if err != nil {
		return nil, fmt.Errorf("bls batch request: %w", err)
	}
	defer func() { _ = resp.Body.Close() }()

	var raw apiResponse
	if err := json.NewDecoder(resp.Body).Decode(&raw); err != nil {
		return nil, fmt.Errorf("decode bls response: %w", err)
	}

	return ParseResponse(raw, meta, s.logger)
}

// ParseResponse turns a decoded BLS response into Observations.
// Exported so unit tests can exercise it without HTTP.
func ParseResponse(raw apiResponse, meta map[string]seriesMeta, logger *slog.Logger) ([]types.Observation, error) {
	if logger == nil {
		logger = pkglog.Default()
	}
	if raw.Status != requestSucceeded {
		msg := strings.Join(raw.Message, "; ")
		if msg == "" {
			msg = "unknown error"
		}
		return nil, fmt.Errorf("bls request failed: %s", msg)
	}

	received := make(map[string]struct{}, len(raw.Results.Series))
	for _, s := range raw.Results.Series {
		received[s.SeriesID] = struct{}{}
	}
	for id := range meta {
		if _, ok := received[id]; !ok {
			logger.Warn("bls: series absent from response", "series", id)
		}
	}

	out := make([]types.Observation, 0)
	for _, s := range raw.Results.Series {
		m, ok := meta[s.SeriesID]
		if !ok {
			logger.Warn("bls: unexpected series in response; skipping", "series", s.SeriesID)
			continue
		}
		count := 0
		for _, p := range s.Data {
			observedAt := PeriodToISO(p.Year, p.Period)
			if observedAt == "" {
				continue
			}
			val, err := strconv.ParseFloat(p.Value, 64)
			if err != nil {
				continue
			}
			out = append(out, types.Observation{
				Source:     "bls",
				Series:     s.SeriesID,
				FIPS:       m.fips,
				Metric:     m.metric,
				ObservedAt: observedAt,
				Value:      val,
				Unit:       m.unit,
			})
			count++
		}
		logger.Info("bls: parsed series", "series", s.SeriesID, "fips", m.fips, "count", count)
	}
	return out, nil
}
